api/handlers.go

```diff
@@ -112,23 +140,106 @@ func (api *API) userHandlerPUT(w http.ResponseWriter, req *http.Request, ps http
+		u, err = api.staticDB.UserByID(req.Context(), id)
+		if err != nil {
+			// This is a Bad Request and not an Internal Server Error because
+			// the client has supplied an invalid user id.
+			WriteError(w, errors.AddContext(err, "failed to fetch user"), http.StatusBadRequest)
+			return
+		}
+	}
+	err = req.ParseMultipartForm(MaxMultipartMem)
+	if err != nil {
+		WriteError(w, errors.New("Failed to parse multipart parameters."), http.StatusBadRequest)
+		return
+	}
+	// Validate that the given old password is correct.
+	if pw := req.PostFormValue("oldPassword"); pw != "" {
+		err = u.VerifyPassword(pw)
+		if err != nil {
+			WriteError(w, errors.New("Bad username or password."), http.StatusBadRequest)
+			return
+		}
+	}
+	// Set the new password.
+	if pw := req.PostFormValue("newPassword"); pw != "" {
+		err = u.SetPassword(pw)
+		if err != nil {
+			WriteError(w, err, http.StatusInternalServerError)
+			return
+		}
+	}
+	// Persist the change.
+	err = api.staticDB.UserUpdatePassword(req.Context(), u)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	WriteJSON(w, u)
 }
 
 // userLoginHandler starts a new session for a user.
 func (api *API) userLoginHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	// TODO Implement
-	fmt.Println("User login.")
-	//req.ParseForm()
-	//req.Form.Set("async", "true")
-	//api.renterDownloadHandler(w, req, ps)
-	WriteJSON(w, struct{ msg string }{"Not implemented."})
+	if err := req.ParseMultipartForm(MaxMultipartMem); err != nil {
+		WriteError(w, errors.New("Failed to parse multipart parameters."), http.StatusBadRequest)
+		return
+	}
+	email, err := database.NewEmail(req.PostFormValue("email"))
+	if err != nil {
+		WriteError(w, database.ErrInvalidEmail, http.StatusBadRequest)
+		return
+	}
+	pw := req.PostFormValue("password")
+	if len(pw) == 0 {
+		WriteError(w, errors.New("The password cannot be empty."), http.StatusBadRequest)
+		return
+	}
+	u, err := api.staticDB.UserByEmail(req.Context(), email)
+	if err != nil {
+		// TODO Consider logging this with a unique ID and returning only the ID
+		// 	and a vague description. Maybe for all endpoints.
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if u.Tier == database.TierUnconfirmed {
+		// TODO This error should be picked up by the FE and it should offer to
+		// 	resend the account confirmation email.
+		WriteError(w, ErrAccountUnconfirmed, http.StatusUnauthorized)
+		return
+	}
+	err = u.VerifyPassword(pw)
+	if err != nil {
+		WriteError(w, errors.New("Bad username or password."), http.StatusUnauthorized)
+		return
+	}
+	token, err := IssueToken(u)
+	if err != nil {
+		fmt.Println(err)
+		// TODO WriteError doesn't set the response's error message properly. Or Postman doesn't read it properly?
+		WriteError(w, err, http.StatusUnprocessableEntity)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	WriteJSON(w, token)
```

Comment:
Something @kwypchlo brought up today was that I can set the user's cookie right here instead of relying on the FE to do that for me. I'll do that in a follow up.